### PR TITLE
[Routing] Optimised dumped router matcher, prevent unneeded function calls.

### PR DESCRIPTION
**Pull request by frankdejonge**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | none
| License       | MIT
| Doc PR        | does not apply

The application I'm working on is fairly large. Because we had a routing issue (not caused by the framework) I looked through the dumped routing code. I spotted some easy wins. These changes brought down the time for the `match` method to run from ~ 7.5ms to ~2.5ms. It's not a lot, but it's something. I've profiled it several times with blackfire to confirm. The results were very consistent. Mind you, our application has quite a serious amount of routes, a little over 900.

### CODE REVIEW

File: src/Symfony/Component/Routing/Matcher/Dumper/PhpMatcherDumper.php [hunk 1]
```diff
@@ -276,7 +278,7 @@ private function compileRoute(Route $route, $name, $supportsRedirections, $paren
             } else {
                 $methods = implode("', '", $methods);
                 $code .= <<<EOF
-            if (!in_array(\$this->context->getMethod(), array('$methods'))) {
```
Comment (frankdejonge) [inline src/Symfony/Component/Routing/Matcher/Dumper/PhpMatcherDumper.php]: This removes N function calls.

File: src/Symfony/Component/Routing/Matcher/Dumper/PhpMatcherDumper.php [hunk 2]
```diff
@@ -266,7 +268,7 @@ private function compileRoute(Route $route, $name, $supportsRedirections, $paren
         if ($methods) {
             if (1 === count($methods)) {
                 $code .= <<<EOF
-            if (\$this->context->getMethod() != '$methods[0]') {
```
Comment (frankdejonge) [inline src/Symfony/Component/Routing/Matcher/Dumper/PhpMatcherDumper.php]: This removes N function calls.

File: src/Symfony/Component/Routing/Matcher/Dumper/PhpMatcherDumper.php [hunk 3]
```diff
@@ -227,7 +229,7 @@ private function compileRoute(Route $route, $name, $supportsRedirections, $paren
 
         if (!count($compiledRoute->getPathVariables()) && false !== preg_match('#^(.)\^(?P<url>.*?)\$\1#'.(substr($regex, -1) === 'u' ? 'u' : ''), $regex, $m)) {
             if ($supportsTrailingSlash && substr($m['url'], -1) === '/') {
-                $conditions[] = sprintf("rtrim(\$pathinfo, '/') === %s", var_export(rtrim(str_replace('\\', '', $m['url']), '/'), true));
```
Comment (frankdejonge) [inline src/Symfony/Component/Routing/Matcher/Dumper/PhpMatcherDumper.php]: This removed N trim calls.

File: src/Symfony/Component/Routing/Matcher/Dumper/PhpMatcherDumper.php [hunk 4]
```diff
@@ -133,7 +135,7 @@ private function compileRoutes(RouteCollection $routes, $supportsRedirections)
         foreach ($groups as $collection) {
             if (null !== $regex = $collection->getAttribute('host_regex')) {
                 if (!$fetchedHost) {
-                    $code .= "        \$host = \$this->context->getHost();\n\n";
```
Comment (frankdejonge) [inline src/Symfony/Component/Routing/Matcher/Dumper/PhpMatcherDumper.php]: This removes a reference lookup, probably hardly any impact, but in optimised code, every cycle counts, right?

File: src/Symfony/Component/Routing/Tests/Fixtures/dumper/url_matcher2.php [hunk 5]
```diff
@@ -35,7 +37,7 @@ public function match($pathinfo)
         if (0 === strpos($pathinfo, '/bar')) {
             // bar
             if (preg_match('#^/bar/(?P<foo>[^/]++)$#s', $pathinfo, $matches)) {
-                if (!in_array($this->context->getMethod(), array('GET', 'HEAD'))) {
+                if (!in_array($requestMethod, array('GET', 'HEAD'))) {
```
Comment (fabpot) [inline src/Symfony/Component/Routing/Tests/Fixtures/dumper/url_matcher2.php]: Reading this line makes me think that we could even have `$requestMethod` set to `GET` when it's `HEAD`, so that this condition can be simplified to just `if ('GET' !== $requestMethod) {` WDTY?

Comment (frankdejonge) [inline src/Symfony/Component/Routing/Tests/Fixtures/dumper/url_matcher2.php]: Wouldn't we then loose the ability to distinguish the request when there's only a HEAD route defined? I see in the there's a case for adding HEAD to the accepted methods when it's a GET request, but not the other way around. So if we only have a HEAD registered, we wouldn't be able to match that anymore, right? This would only be the case if you want to support defining separate HEAD handlers, I don't know if that's currently the case.

Comment (frankdejonge) [inline src/Symfony/Component/Routing/Tests/Fixtures/dumper/url_matcher2.php]: I'll just test this out real quick.

Comment (frankdejonge) [inline src/Symfony/Component/Routing/Tests/Fixtures/dumper/url_matcher2.php]: I just tested this out. We'd loose the ability to define HEAD routes when we make that additional optimisation.

Comment (fabpot) [inline src/Symfony/Component/Routing/Tests/Fixtures/dumper/url_matcher2.php]: Sorry, I wasn't clear enough. My idea was to create a specific variable (different from `$requestMethod`, like `$isLikeGetMethod` or something) that holds `GET` when the real method is `HEAD`.

Comment (frankdejonge) [inline src/Symfony/Component/Routing/Tests/Fixtures/dumper/url_matcher2.php]: That shaved off a `0.2ms`, which is not a lot. Was done in this commit: 0425f33

Comment (frankdejonge) [inline src/Symfony/Component/Routing/Tests/Fixtures/dumper/url_matcher2.php]: I've implemented it in such a way that it keeps current behaviour, which also became more clear to me. If there's a HEAD registered after a GET for the same route, the first route will be matched. So now there is also a variable for the special "GET is also HEAD" case. The compiled route needs to remain aware whether there was actually a HEAD clause in there initially so it uses the right request method variable, the not "special" one.

Comment (frankdejonge) [inline src/Symfony/Component/Routing/Tests/Fixtures/dumper/url_matcher2.php]: I begin to wonder though, the code generating this, and executing it, become less easy to understand. Since the boost is so enormously tiny of that last bit I suggest to consider reverting that one commit and keep the rest.

Comment (frankdejonge) [inline src/Symfony/Component/Routing/Tests/Fixtures/dumper/url_matcher2.php]: I've added the tests for the other optimisation just in case. It's 03:00 here now, so bed time. Have a good one 👍

File: src/Symfony/Component/Routing/Tests/Fixtures/dumper/url_matcher1.php [hunk 6]
```diff
@@ -24,8 +24,15 @@ public function match($pathinfo)
     {
         $allow = array();
         $pathinfo = rawurldecode($pathinfo);
+        $trimmedPathinfo = rtrim($pathinfo, '/');
         $context = $this->context;
         $request = $this->request;
+        $requestMethod = $isLikeGetMethod = $context->getMethod();
+
+        if ($requestMethod === 'HEAD') {
```
Comment (hhamon) [inline src/Symfony/Component/Routing/Tests/Fixtures/dumper/url_matcher1.php]: Should be `if ('HEAD' === $requestMethod) {`

File: src/Symfony/Component/Routing/Tests/Fixtures/dumper/url_matcher1.php [hunk 7]
```diff
@@ -46,8 +53,8 @@ public function match($pathinfo)
 
             // barhead
             if (0 === strpos($pathinfo, '/barhead') && preg_match('#^/barhead/(?P<foo>[^/]++)$#s', $pathinfo, $matches)) {
-                if (!in_array($this->context->getMethod(), array('GET', 'HEAD'))) {
-                    $allow = array_merge($allow, array('GET', 'HEAD'));
+                if ($isLikeGetMethod != 'GET') {
```
Comment (hhamon) [inline src/Symfony/Component/Routing/Tests/Fixtures/dumper/url_matcher1.php]: `if ('GET' !== $isLikeGetMethod) {`

File: src/Symfony/Component/Routing/Tests/Fixtures/dumper/url_matcher1.php [hunk 8]
```diff
@@ -83,7 +90,7 @@ public function match($pathinfo)
 
             // baz5
             if (preg_match('#^/test/(?P<foo>[^/]++)/$#s', $pathinfo, $matches)) {
-                if ($this->context->getMethod() != 'POST') {
+                if ($isLikeGetMethod != 'POST') {
```
Comment (hhamon) [inline src/Symfony/Component/Routing/Tests/Fixtures/dumper/url_matcher1.php]: `if ('POST' !== $isLikeGetMethod) {`

File: src/Symfony/Component/Routing/Tests/Fixtures/dumper/url_matcher1.php [hunk 9]
```diff
@@ -94,7 +101,7 @@ public function match($pathinfo)
 
             // baz.baz6
             if (preg_match('#^/test/(?P<foo>[^/]++)/$#s', $pathinfo, $matches)) {
-                if ($this->context->getMethod() != 'PUT') {
+                if ($isLikeGetMethod != 'PUT') {
```
Comment (hhamon) [inline src/Symfony/Component/Routing/Tests/Fixtures/dumper/url_matcher1.php]: `if ('PUT' !== $isLikeGetMethod) {`

File: src/Symfony/Component/Routing/Tests/Fixtures/dumper/url_matcher2.php [hunk 10]
```diff
@@ -24,8 +24,15 @@ public function match($pathinfo)
     {
         $allow = array();
         $pathinfo = rawurldecode($pathinfo);
+        $trimmedPathinfo = rtrim($pathinfo, '/');
         $context = $this->context;
         $request = $this->request;
+        $requestMethod = $isLikeGetMethod = $context->getMethod();
+
+        if ($requestMethod === 'HEAD') {
```
Comment (hhamon) [inline src/Symfony/Component/Routing/Tests/Fixtures/dumper/url_matcher2.php]: `if ('HEAD' !== $isLikeGetMethod) {`

File: src/Symfony/Component/Routing/Tests/Fixtures/dumper/url_matcher2.php [hunk 11]
```diff
@@ -46,8 +53,8 @@ public function match($pathinfo)
 
             // barhead
             if (0 === strpos($pathinfo, '/barhead') && preg_match('#^/barhead/(?P<foo>[^/]++)$#s', $pathinfo, $matches)) {
-                if (!in_array($this->context->getMethod(), array('GET', 'HEAD'))) {
-                    $allow = array_merge($allow, array('GET', 'HEAD'));
+                if ($isLikeGetMethod != 'GET') {
```
Comment (hhamon) [inline src/Symfony/Component/Routing/Tests/Fixtures/dumper/url_matcher2.php]: `if ('GET' !== $isLikeGetMethod) {`

File: src/Symfony/Component/Routing/Tests/Fixtures/dumper/url_matcher2.php [hunk 12]
```diff
@@ -91,7 +98,7 @@ public function match($pathinfo)
 
             // baz5
             if (preg_match('#^/test/(?P<foo>[^/]++)/$#s', $pathinfo, $matches)) {
-                if ($this->context->getMethod() != 'POST') {
+                if ($isLikeGetMethod != 'POST') {
```
Comment (hhamon) [inline src/Symfony/Component/Routing/Tests/Fixtures/dumper/url_matcher2.php]: `if ('POST' !== $isLikeGetMethod) {`

File: src/Symfony/Component/Routing/Tests/Fixtures/dumper/url_matcher2.php [hunk 13]
```diff
@@ -102,7 +109,7 @@ public function match($pathinfo)
 
             // baz.baz6
             if (preg_match('#^/test/(?P<foo>[^/]++)/$#s', $pathinfo, $matches)) {
-                if ($this->context->getMethod() != 'PUT') {
+                if ($isLikeGetMethod != 'PUT') {
```
Comment (hhamon) [inline src/Symfony/Component/Routing/Tests/Fixtures/dumper/url_matcher2.php]: `if ('PUT' !== $isLikeGetMethod) {`

**Comment by frankdejonge** [comment]
@hhamon I get it, joda-ing it now :P

File: src/Symfony/Component/Routing/Tests/Fixtures/dumper/url_matcher2.php [hunk 14]
```diff
@@ -174,7 +181,7 @@ public function match($pathinfo)
             }
 
             // hey
-            if (rtrim($pathinfo, '/') === '/multi/hey') {
+            if ($trimmedPathinfo === '/multi/hey') {
```
Comment (hhamon) [inline src/Symfony/Component/Routing/Tests/Fixtures/dumper/url_matcher2.php]: `if ('/multi/hey' === $trimmedPathinfo) {`

File: src/Symfony/Component/Routing/Tests/Fixtures/dumper/url_matcher3.php [hunk 15]
```diff
@@ -24,8 +24,15 @@ public function match($pathinfo)
     {
         $allow = array();
         $pathinfo = rawurldecode($pathinfo);
+        $trimmedPathinfo = rtrim($pathinfo, '/');
         $context = $this->context;
         $request = $this->request;
+        $requestMethod = $isLikeGetMethod = $context->getMethod();
+
+        if ($requestMethod === 'HEAD') {
```
Comment (hhamon) [inline src/Symfony/Component/Routing/Tests/Fixtures/dumper/url_matcher3.php]: `if ('HEAD' === $requestMethod) {`

File: src/Symfony/Component/Routing/Matcher/Dumper/PhpMatcherDumper.php [hunk 16]
```diff
@@ -105,8 +105,15 @@ public function match(\$pathinfo)
     {
         \$allow = array();
         \$pathinfo = rawurldecode(\$pathinfo);
+        \$trimmedPathinfo = rtrim(\$pathinfo, '/');
         \$context = \$this->context;
         \$request = \$this->request;
+        \$requestMethod = \$isLikeGetMethod = \$context->getMethod();
+
+        if (\$requestMethod === 'HEAD') {
+            \$isLikeGetMethod = 'GET';
```
Comment (stof) [inline src/Symfony/Component/Routing/Matcher/Dumper/PhpMatcherDumper.php]: the ``$isLikeGetMethod`` variable name makes me think it is a boolean

**Comment by frankdejonge** [comment]
@hhamon I've corrected the non-yoda to yoda statement, I've also corrected the ones which I didn't change into yoda statements.

File: src/Symfony/Component/Routing/Tests/Fixtures/dumper/url_matcher2.php [hunk 17]
```diff
@@ -35,8 +42,8 @@ public function match($pathinfo)
         if (0 === strpos($pathinfo, '/bar')) {
             // bar
             if (preg_match('#^/bar/(?P<foo>[^/]++)$#s', $pathinfo, $matches)) {
-                if (!in_array($this->context->getMethod(), array('GET', 'HEAD'))) {
-                    $allow = array_merge($allow, array('GET', 'HEAD'));
+                if (!in_array($isLikeGetMethod, array('GET'))) {
+                    $allow = array_merge($allow, array('GET'));
```
Comment (stof) [inline src/Symfony/Component/Routing/Tests/Fixtures/dumper/url_matcher2.php]: not adding HEAD anymore to the allow array looks suspicious to me.

Comment (frankdejonge) [inline src/Symfony/Component/Routing/Tests/Fixtures/dumper/url_matcher2.php]: @stof I'm open to suggestions here. The name was given by @fabpot initially, I couldn't come up with a better name for it now. The name goes into the realm of "normalised" pretty quickly which doesn't add a lot of context.

Comment (stof) [inline src/Symfony/Component/Routing/Tests/Fixtures/dumper/url_matcher2.php]: @frankdejonge are you sure you replied to the right comment ? This looks unrelated

**Comment by frankdejonge** [comment]
@stof I've corrected the case where only one method is present after filtering out the HEAD method.

File: src/Symfony/Component/Routing/Tests/Fixtures/dumper/url_matcher2.php (hunk 17 above)
Comment (frankdejonge) [inline src/Symfony/Component/Routing/Tests/Fixtures/dumper/url_matcher2.php]: @stof I think github must have durped, because I looked at it when writing but now it's placed here. Super weird.

File: src/Symfony/Component/Routing/Tests/Fixtures/dumper/url_matcher1.php [hunk 18]
```diff
@@ -35,8 +42,8 @@ public function match($pathinfo)
         if (0 === strpos($pathinfo, '/bar')) {
             // bar
             if (preg_match('#^/bar/(?P<foo>[^/]++)$#s', $pathinfo, $matches)) {
-                if (!in_array($this->context->getMethod(), array('GET', 'HEAD'))) {
-                    $allow = array_merge($allow, array('GET', 'HEAD'));
+                if ('GET' !== $isLikeGetMethod) {
+                    $allow[] = 'GET';
```
Comment (stof) [inline src/Symfony/Component/Routing/Tests/Fixtures/dumper/url_matcher1.php]: Why aren't you adding HEAD anymore here ? It is an allowed method

Comment (frankdejonge) [inline src/Symfony/Component/Routing/Tests/Fixtures/dumper/url_matcher1.php]: @stof this was an optimisation proposed by @fabpot. We stream HEAD in a special way (this is current behaviour). When a HEAD request is made, we also match GET. Inverted, when we declare GET actions, we also match in HEAD. So in this normalised case we know that if something is "like get" we don't need HEAD anymore because we're we already match GET.

**Comment by frankdejonge** [comment]
Perhaps it would be good to rename `isGetLikeMethod` to `methodWhereHeadMatchesGet`? Might be a bit verbose, but in this case I think being explicit is preferable. WDYT @stof?

**Comment by frankdejonge** [comment]
I've also added another test fixture which more clearly demonstrates the effect of the method optimisation.

File: src/Symfony/Component/Routing/Matcher/Dumper/PhpMatcherDumper.php [hunk 19]
```diff
@@ -105,8 +105,16 @@ public function match(\$pathinfo)
     {
         \$allow = array();
         \$pathinfo = rawurldecode(\$pathinfo);
+        \$trimmedPathinfo = rtrim(\$pathinfo, '/');
         \$context = \$this->context;
         \$request = \$this->request;
+        \$requestMethod = \$isLikeGetMethod = \$context->getMethod();
+        \$schema = \$context->getScheme();
+
+        if (\$requestMethod === 'HEAD') {
```
Comment (GuilhemN) [inline src/Symfony/Component/Routing/Matcher/Dumper/PhpMatcherDumper.php]: `'HEAD' ===` 😉

Comment (frankdejonge) [inline src/Symfony/Component/Routing/Matcher/Dumper/PhpMatcherDumper.php]: Fixed, thanks!

File: src/Symfony/Component/Routing/Tests/Fixtures/dumper/url_matcher1.php (hunk 18 above)
Comment (Jean85) [inline src/Symfony/Component/Routing/Tests/Fixtures/dumper/url_matcher1.php]: Can we make sure that this is properly tested? If someone else breaks this, it will be a huge bug!

File: src/Symfony/Component/Routing/Matcher/Dumper/PhpMatcherDumper.php [hunk 20]
```diff
@@ -105,8 +105,16 @@ public function match(\$pathinfo)
...
         \$allow = array();
         \$pathinfo = rawurldecode(\$pathinfo);
+        \$trimmedPathinfo = rtrim(\$pathinfo, '/');
         \$context = \$this->context;
         \$request = \$this->request;
+        \$requestMethod = \$isLikeGetMethod = \$context->getMethod();
+        \$schema = \$context->getScheme();
+
+        if ('HEAD' === \$requestMethod) {
+            \$isLikeGetMethod = 'GET';
```
Comment (javiereguiluz) [inline src/Symfony/Component/Routing/Matcher/Dumper/PhpMatcherDumper.php]: The `is` prefix in this variable name looks wrong. Maybe it's too late, but what if we rename `$isLikeGetMethod = 'GET'` to `$isSafeMethod = true`

We had some discussions about what a "safe method" is in HTTP, but the section 9.1.1 of RFC 2616 mentions `GET` and `HEAD` as safe methods: http://www.w3.org/Protocols/rfc2616/rfc2616-sec9.html

File: src/Symfony/Component/Routing/Tests/Fixtures/dumper/url_matcher1.php (hunk 18 above)
Comment (frankdejonge) [inline src/Symfony/Component/Routing/Tests/Fixtures/dumper/url_matcher1.php]: @Jean85 there's an additional test case added in this PR to do just that.

File: src/Symfony/Component/Routing/Matcher/Dumper/PhpMatcherDumper.php [hunk 21]
```diff
@@ -263,26 +266,57 @@ private function compileRoute(Route $route, $name, $supportsRedirections, $paren
...
+                    // Since we treat HEAD requests like GET requests we don't need to match it.
+                    $methodVariable = 'isLikeGetMethod';
+                    $methods = array_filter($methods, function ($method) { return 'HEAD' !== $method; });
+                }
+
+                if (1 === count($methods)) {
+                    $code .= <<<EOF
+            if ('$methods[0]' !== \$$methodVariable) {
+                \$allow[] = '$methods[0]';
+                goto $gotoname;
```
Comment (GuilhemN) [inline src/Symfony/Component/Routing/Matcher/Dumper/PhpMatcherDumper.php]: Was the use of the goto ever discussed?
It seems like here we only need to reverse the conditions to get ride of it.

File: src/Symfony/Component/Routing/Matcher/Dumper/PhpMatcherDumper.php (hunk 20 above)
Comment (frankdejonge) [inline src/Symfony/Component/Routing/Matcher/Dumper/PhpMatcherDumper.php]: Actually, we don't want a boolean, because then we can't perform a meaningful optimisation. We just want to have a interpreted method, which we can check against. Otherwise we'd need to do more check than less.

**Review by Jean85 — APPROVED** [review]

**Comment by frankdejonge** [comment]
Tests currently seem to be failing because master is broken.

File: src/Symfony/Component/Routing/Matcher/Dumper/PhpMatcherDumper.php (hunk 20 above)
Comment (javiereguiluz) [inline src/Symfony/Component/Routing/Matcher/Dumper/PhpMatcherDumper.php]: What about renaming it to `$methodAlias` or `$canonicalMethod` or something like that? We "can't" store a string in a `$is*` variable.

Comment (frankdejonge) [inline src/Symfony/Component/Routing/Matcher/Dumper/PhpMatcherDumper.php]: I like `conicalMethod`, a lot. I had previously opted to use `methodWhereHeadMatchesGet` instead. Which is a little verbose, but it does make it clear what's going on.

Comment (fabpot) [inline src/Symfony/Component/Routing/Matcher/Dumper/PhpMatcherDumper.php]: +1 for `$canonicalMethod`

File: src/Symfony/Component/Routing/Matcher/Dumper/PhpMatcherDumper.php [hunk 22]
```diff
@@ -105,8 +105,16 @@ public function match(\$pathinfo)
     {
         \$allow = array();
         \$pathinfo = rawurldecode(\$pathinfo);
+        \$trimmedPathinfo = rtrim(\$pathinfo, '/');
         \$context = \$this->context;
         \$request = \$this->request;
+        \$requestMethod = \$canonicalMethod = \$context->getMethod();
+        \$schema = \$context->getScheme();
```
Comment (fabpot) [inline src/Symfony/Component/Routing/Matcher/Dumper/PhpMatcherDumper.php]: should be `$scheme` here, not `$schema`, right?

Comment (frankdejonge) [inline src/Symfony/Component/Routing/Matcher/Dumper/PhpMatcherDumper.php]: @fabpot correct, I've fixed it.

**Comment by fabpot** [comment]
👍

**Comment by fabpot** [comment]
Thank you @frankdejonge.

File: src/Symfony/Component/Routing/Matcher/Dumper/PhpMatcherDumper.php (hunk 21 above)
Comment (frankdejonge) [inline src/Symfony/Component/Routing/Matcher/Dumper/PhpMatcherDumper.php]: I'm checking this out now, if proven to be effective I'll create another PR.

Comment (frankdejonge) [inline src/Symfony/Component/Routing/Matcher/Dumper/PhpMatcherDumper.php]: @GuilhemN I suspected it to have some effect, but it did literally nothing.
